src/Functions/FunctionsConversion.h

```diff
@@ -387,125 +389,206 @@ struct ConvertImpl
-    : DateTimeTransformImpl<DataTypeDateTime, DataTypeDate, ToDateImpl> {};
+template <typename Name, DT_OVERFLOW_MODE date_time_overflow_mode>
+struct ConvertImpl<DataTypeDateTime, DataTypeDate, Name, ConvertDefaultBehaviorTag, date_time_overflow_mode>
+    : DateTimeTransformImpl<DataTypeDateTime, DataTypeDate, ToDateImpl<date_time_overflow_mode>, false> {};
 
 /** Conversion of DateTime to Date32: throw off time component.
   */
-template <typename Name> struct ConvertImpl<DataTypeDateTime, DataTypeDate32, Name, ConvertDefaultBehaviorTag>
-    : DateTimeTransformImpl<DataTypeDateTime, DataTypeDate32, ToDate32Impl> {};
+template <typename Name, DT_OVERFLOW_MODE date_time_overflow_mode>
+struct ConvertImpl<DataTypeDateTime, DataTypeDate32, Name, ConvertDefaultBehaviorTag, date_time_overflow_mode>
+    : DateTimeTransformImpl<DataTypeDateTime, DataTypeDate32, ToDate32Impl<date_time_overflow_mode>, false> {};
 
 /** Conversion of Date to DateTime: adding 00:00:00 time component.
   */
+template <DT_OVERFLOW_MODE date_time_overflow_mode = default_date_time_overflow_mode>
 struct ToDateTimeImpl
 {
     static constexpr auto name = "toDateTime";
 
     static UInt32 execute(UInt16 d, const DateLUTImpl & time_zone)
     {
+        if constexpr (date_time_overflow_mode == DT_OVERFLOW_MODE::Throw)
+        {
+            if (d > MAX_DT_DAY_NUM)
+                throw Exception(ErrorCodes::VALUE_IS_OUT_OF_RANGE_OF_DATA_TYPE, "Day number {} is out of bounds of type DateTime", d);
+        }
+        else if constexpr (date_time_overflow_mode == DT_OVERFLOW_MODE::Saturate)
+        {
+            if (d > MAX_DT_DAY_NUM)
+                d = MAX_DT_DAY_NUM;
+        }
         return static_cast<UInt32>(time_zone.fromDayNum(DayNum(d)));
     }
 
-    static Int64 execute(Int32 d, const DateLUTImpl & time_zone)
+    static UInt32 execute(Int32 d, const DateLUTImpl & time_zone)
     {
-        return time_zone.fromDayNum(ExtendedDayNum(d));
+        if constexpr (date_time_overflow_mode == DT_OVERFLOW_MODE::Saturate)
+        {
+            if (d < 0)
+                return 0;
+            else if (d > MAX_DT_DAY_NUM)
+                d = MAX_DT_DAY_NUM;
+        }
+        else if constexpr (date_time_overflow_mode == DT_OVERFLOW_MODE::Throw)
+        {
+            if (d < 0 || d > MAX_DT_DAY_NUM)
+                throw Exception(ErrorCodes::VALUE_IS_OUT_OF_RANGE_OF_DATA_TYPE, "Value {} is out of bounds of type DateTime", d);
+        }
+        return static_cast<UInt32>(time_zone.fromDayNum(ExtendedDayNum(d)));
     }
 
     static UInt32 execute(UInt32 dt, const DateLUTImpl & /*time_zone*/)
     {
         return dt;
     }
 
-    // TODO: return UInt32 ???
-    static Int64 execute(Int64 dt64, const DateLUTImpl & /*time_zone*/)
+    static UInt32 execute(Int64 dt64, const DateLUTImpl & /*time_zone*/)
     {
-        return dt64;
+        if constexpr (date_time_overflow_mode == DT_OVERFLOW_MODE::Ignore)
```

Comment:
Maybe it is a matter of taste, but the code would be easier to read and equally fast with `if constexpr` to cover all cases at the function-level (instead of mixing `if constexpr` and logic)

Reply:
well, it is really a matter of taste 😄 
I like it as it is -- it has minimum amount of duplicated code (only when really necessary)
Otherwise we'll have e.g. lots of similar returns